### ISSUE TITLE
Refactor K8s watcher to use the new subscriber package

### ISSUE THREAD

**Issue by christarazi** [cilium/cilium#15471]
In the v1.10 release cycle, we introduced https://github.com/cilium/cilium/pull/15295 to help with https://github.com/cilium/cilium/pull/15340. However, in this release cycle, we also had https://github.com/cilium/cilium/pull/15134 and https://github.com/cilium/cilium/pull/15383, both of which added onto the K8s watcher pattern that https://github.com/cilium/cilium/pull/15295 was supposed to eliminate. 

Both the agent and the operator have K8s watchers, which means the refactor should encompass both. The operator is much easier to refactor due to less reliance on the service cache. That's what took place in https://github.com/cilium/cilium/pull/15295. However, the agent on the other hand, has many consumers of the service cache, which means refactoring out is a much taller task than converting other watchers to use the new package. So it was left out and this issue opened so that this can be tackled after the release.

Specifically, here's what needs to be done to refactor the K8s watcher in the agent, specifically the service watcher:

When we get a service event, we first push it into our `ServiceCache`. However, that process involves converting the service object (`slim_corev1.Service`) into an internal representation (`pkg/k8s.Service`). This complicates the refactor because  https://github.com/cilium/cilium/pull/15295 assumes that all subscribers will be notified with the original service object, not the internally converted one. In the agent, the redirect policy manager and others rely on the service cache, which in turns means it relies on the converted object. One would need to untangle all dependencies on the `ServiceCache`, and introduce a mapping from original service object -> converted internal representation. Then have a getter to get the converted object from the original. Then the subscribers will look like this when notified, or specifically, the redirect policy manager will have a method like:

```golang
// OnAddService adheres to the subscriber.ServiceHandler interface
func (s *Something) OnAddService(svc *slim_corev1.Service) {
  converted := serviceCache.Get(svc)
  s.doStuffWithServiceOnAdd(converted)
  ...
}
```

There are also open questions around the `lock.StoppableWaitGroup` which is used extensively in the service watcher. It is [created](https://github.com/cilium/cilium/blob/10c301d4d55f2142f0f684bf6f8d6fa84cb9fb55/pkg/k8s/watchers/watcher.go#L357) before the service watcher is initialized and is passed down the stack, all the way to the individual subscribers. Does this object need to exist at this "global" level. Shouldn't it be a part of the subscriber itself? That's what https://github.com/cilium/cilium/pull/15295 did with operator service watcher.

There are other watchers that need to be converted as well in the agent like the pod watcher.

**Comment by brb** [comment]
> When we get a service event, we first push it into our ServiceCache.

I am wondering whether we could kill `ServiceCache`. I'd like to remove yet another level of indirection which `ServiceCache` introduces. E.g. instead of notifying `pkg/service` directly via the k8s service watcher, we do that via the mechanism exposed by `ServiceCache`. This complicates objects synchronization too.

Currently, the main purpose of `ServiceCache` is to corelate Endpoint objects with a `Service` object. If we assume that the consumers of the `ServiceCache` (`pkg/service`, LRP, cluster-mesh, what else?) could deal with incremental changes, then this corelation would be redundant.

**Comment by ldelossa** [comment]
As a note, via discussions on: https://github.com/cilium/cilium/pull/16525/files @christarazi and I have agreed upon a few small changes to the new subscriber patterns. 

Most noteably: 
https://github.com/cilium/cilium/pull/16525#discussion_r653889857
https://github.com/cilium/cilium/pull/16525#discussion_r654401847

@christarazi I think after podCidr work is complete I'll go back and update the existing subscriber.Service interfaces and implementations to match the changes discussed in the above comments.

**Comment by MikeLing** [comment]
Hi @christarazi , could I work on this issue? Could give me some clues about where should I start? Thanks

**Comment by aanm** [comment]
Hi @MikeLing thanks for volunteering, this issue is already being taking care partially by @joamaki as he is performing a heavy refactoring in the core-code that will allow to have a subscriber package afterwards.

**Comment by joamaki** [comment]
This is slowly progressing with `pkg/k8s/resource` utility merged and #21744 switching service, node and namespace watchers over to it.

There are few complications slowing this work down:
- Some watchers are flipping to different field selectors at runtime (pod watcher will start watching only local node when kvstore connects). Not yet clear how a `Resource[*Pod]` should work. One option is to allow changing the `ListerWatcher` after starting, but this has nasty semantic issues, e.g. store would get suddenly cleared and repopulated and we'd have another sync event. Not sure how event subscribers should handle this. Likely we should not expose `Resource[*Pod]` directly to all modules, but rather something that hides whether pod-related events are coming from k8s or kvstore.
- With pkg/k8s/resource we would no longer have the global "all resources has synced and handlers are done signal" (`Resource.WaitForCacheSync`). We'll need a more explicit signalling for readiness instead that is not tied to the specific source of events. I have some ideas how to go about this, but it'll definitely not make it into 1.13.
